File: client/swift/Sources/WasmRpc/Transport.swift
import Foundation

/// Moves opaque protobuf frames to a wasm-rpc router.
public protocol WasmRpcTransport: Sendable {
    func invoke(method: String, payload: Data) async throws -> Data
}

/// Server-streaming extension of ``WasmRpcTransport``.
public protocol WasmRpcStreamTransport: WasmRpcTransport {
    /// Yields one `Data` frame per server message; finishes on end,
    /// throws ``WasmRpcError`` on error. Terminating the sequence
    /// cancels the underlying stream.
    func listen(method: String, payload: Data) -> AsyncThrowingStream<Data, Error>
}

/// Standardized rejection payload produced by the Go bridge.
public struct WasmRpcError: Error, Sendable {
    public let code: String
    public let method: String?
    public let message: String

    public init(code: String, method: String? = nil, message: String) {
        self.code = code
        self.method = method
        self.message = message
    }
}

File: client/kotlin/src/main/kotlin/dev/hyperauth/wasmrpc/Transport.kt
package dev.hyperauth.wasmrpc

import kotlinx.coroutines.flow.Flow

/** Moves opaque protobuf frames to a wasm-rpc router. */
interface WasmRpcTransport {
    suspend fun invoke(method: String, payload: ByteArray): ByteArray
}

/** Server-streaming extension of [WasmRpcTransport]. */
interface WasmRpcStreamTransport : WasmRpcTransport {
    /**
     * Emits one frame per server message; completes on end, throws
     * [WasmRpcException] on error. Collector cancellation cancels the
     * underlying stream.
     */
    fun listen(method: String, payload: ByteArray): Flow<ByteArray>
}

/** Standardized rejection payload produced by the Go bridge. */
class WasmRpcException(
    val code: String,
    val method: String?,
    message: String,
) : Exception("$code: $message")
